plugin-server/src/main/ingestion-queues/batch-processing/each-batch-buffer.ts

```diff
@@ -19,42 +18,47 @@ export async function eachMessageBuffer(
         timeoutMessage: 'After 30 seconds still running runBufferEventPipeline',
     })
     resolveOffset(message.offset)
+    await heartbeat()
 }
 
 export async function eachBatchBuffer(
-    { batch, resolveOffset, commitOffsetsIfNecessary }: EachBatchPayload,
+    { batch, resolveOffset, heartbeat, commitOffsetsIfNecessary }: EachBatchPayload,
     queue: KafkaQueue
 ): Promise<void> {
     if (batch.messages.length === 0) {
         return
     }
     const batchStartTimer = new Date()
 
+    /** Index of first message to be processed post-sleep. -1 means there's no sleep needed. */
+    let messagesCutoffIndex = -1
+    /** How long should we sleep for until we have a desired delay in processing. */
     let consumerSleepMs = 0
-    for (const message of batch.messages) {
-        // kafka timestamps are unix timestamps in string format
+    for (let i = 0; i < batch.messages.length; i++) {
+        const message = batch.messages[i]
+        // Kafka timestamps are Unix timestamps in string format
         const processAt = Number(message.timestamp) + queue.pluginsServer.BUFFER_CONVERSION_SECONDS * 1000
         const delayUntilTimeToProcess = processAt - Date.now()
 
-        if (delayUntilTimeToProcess < 0) {
-            await eachMessageBuffer(message, resolveOffset, queue)
+        if (delayUntilTimeToProcess <= 0 && messagesCutoffIndex < 0) {
+            await eachMessageBuffer(message, resolveOffset, heartbeat, queue)
         } else {
-            consumerSleepMs = Math.max(consumerSleepMs, delayUntilTimeToProcess)
+            if (messagesCutoffIndex < 0) {
+                messagesCutoffIndex = i
+            }
+            if (delayUntilTimeToProcess > consumerSleepMs) {
+                consumerSleepMs = delayUntilTimeToProcess
+            }
         }
     }
-
-    // if consumerSleep > 0 it means we didn't process at least one message
-    if (consumerSleepMs > 0) {
-        // pause the consumer for this partition until we can process all unprocessed messages from this batch
-        await queue.bufferSleep(consumerSleepMs, batch.partition)
-
-        // we throw an error to prevent the non-processed message offsets from being committed
-        // from the kafkajs docs:
-        // > resolveOffset() is used to mark a message in the batch as processed.
-        // > In case of errors, the consumer will automatically commit the resolved offsets.
-        throw new DelayProcessing()
+    await commitOffsetsIfNecessary()
+    if (messagesCutoffIndex >= 0) {
+        // Pause the consumer for this partition until we can process all unprocessed messages from this batch
+        await queue.bufferSleep(consumerSleepMs, batch.partition, heartbeat)
+        for (const message of batch.messages.slice(messagesCutoffIndex)) {
+            await eachMessageBuffer(message, resolveOffset, heartbeat, queue)
+        }
```

Comment:
This will only pause the `{ topic, partition }` combination though. [KafkaJS docs](https://kafka.js.org/docs/consuming#a-name-pause-resume-a-pause-resume) Therefore this _shouldn't_ be an issue, unless I'm missing something about KafkaJS.

Reply:
The problem isn't what we pause - it's that we won't have marked this batch as finished. I'm reasonably sure that even with the pause, if `partitionsConsumedConcurrently` is 1, we'll not process any other partition before this batch completes

Reply:
That is a good point. I thought about this more and implemented a solution suggested in https://github.com/tulios/kafkajs/issues/612 – this way the batch is finished as soon as we reach the cutoff message (i.e. the last one which is delayed enough) + the partition is `.pause()`d. After the timeout we `.seek()` to the cutoff message and only then do we `.resume()` the partition, which ensures all message get processed.

Reply:
Had to upgrade to upgrade to KafkaJS 2.0.2 as `.seek()` is broken in 2.0.1 (https://github.com/tulios/kafkajs/releases/tag/v2.0.2)